pallene/translator.lua

```diff
@@ -31,77 +31,124 @@ local Translator = util.Class()
 end
 
+function Translator:add_local(start_index, stop_index)
+    self:add_previous(start_index - 1)
+    -- The export keyword is six characters long, whereas the local keyword is five characters
+    -- long. Therefore, we pad the keyword with a space. We could add the space to the left, too.
+    -- But it seems more "natural" on the right.
+    table.insert(self.partials, "local ")
+
+    self.last_index = stop_index
+end
+
+function Translator:add_exports()
+    if #self.exports > 0 then
+        table.insert(self.partials, "\nreturn {\n")
+        for _, export in ipairs(self.exports) do
+            local pair = string.format("    %s = %s,\n", export, export)
+            table.insert(self.partials, pair)
+        end
+        table.insert(self.partials, "}\n")
+    end
+end
+
+function Translator:translate_var_decl(decl)
+    if decl.type then
+        -- Remove the colon but retain any adjacent comment to the right.
+        self:add_whitespace(decl.col_loc.pos, decl.col_loc.pos)
+        -- Remove the type annotation but exclude the next token.
+        self:add_whitespace(decl.type.loc.pos, decl.end_loc.pos - 1)
+    end
+end
+
+function Translator:translate_tl_variables(node)
+    -- Add the variables to the export sequence if they are declared with the `export`
+    -- modifier.
+    if not node.is_local then
+        self:add_local(node.loc.pos, node.mod_end_loc.pos - 1)
+        for _, decl in ipairs(node.decls) do
+            table.insert(self.exports, decl.name)
+        end
+    end
+
+    for _, decl in ipairs(node.decls) do
+        self:translate_var_decl(decl)
+    end
+end
+
+function Translator:translate_block(block)
+    for _, statement in ipairs(block.stats) do
+        if statement._tag == "ast.Stat.Decl" then
+            for _, decl in ipairs(statement.decls) do
+                self:translate_var_decl(decl)
+            end
+        elseif statement._tag == "ast.Stat.For" then
+            self:translate_var_decl(statement.decl)
+        elseif statement._tag == "ast.Stat.Block" then
+            self:translate_block(statement)
+        end
+    end
+end
+
+function Translator:translate_tl_function(node)
+    if not node.is_local then
+        self:add_local(node.loc.pos, node.mod_end_loc.pos - 1)
+        table.insert(self.exports, node.decl.name)
+    end
+
+    -- Remove type annotations from function parameters.
+    for _, arg_decl in ipairs(node.value.arg_decls) do
+        self:translate_var_decl(arg_decl)
+    end
+
+    -- Remove type annotations from the return type, which is optional. However, `rt_col_loc`
+    -- and `rt_end_loc` are always set. Therefore, it is safe to replace without any checks.
+    self:add_whitespace(node.rt_col_loc.pos, node.rt_end_loc.pos - 1)
+
+    self:translate_block(node.value.body)
+end
+
 function translator.translate(input, prog_ast)
     local instance = Translator:new()
-
-    print(require("inspect")(prog_ast))
+    instance.input = input
```

Comment:
Two things here:

1. We should be using `Translator.new()` instead of `Translator:new()`. Our util.class expects the former.
2. We could pass the input as a parameter to the constructor instead of initializing it separately.

Fixing problem number 1, which we had overlooked before, is a prerequisite for fixing problem number 2.